employee/read_only.c

```diff
@@ -32,11 +34,13 @@ int main()
  */
 struct employee *set_read_only(struct employee *e)
 {
-	char *ro_name = (char *)cheri_perms_and(e->name, CHERI_PERM_LOAD);
+	char *ro_alpha_id = (char *)cheri_perms_and(e->alpha_id, CHERI_PERM_LOAD);
+	char *ro_name = (char *)cheri_perms_and(e->name,  CHERI_PERM_LOAD);
 	char *ro_surname = (char *)cheri_perms_and(e->surname, CHERI_PERM_LOAD);
-	struct employee *ro_employee =
-		(struct employee *)cheri_perms_and(e, CHERI_PERM_LOAD || CHERI_PERM_LOAD_CAP);
-	ro_employee->name = ro_name;
-	ro_employee->surname = ro_surname;
-	return ro_employee;
+	struct employee ro_employee = {ro_alpha_id, ro_name, ro_surname, e->salary};
+    struct employee *ro_employee_ptr = &ro_employee;
```

Comment:
This returns a dangling pointer, since `ro_employee` is on the local stack. I suspect that this is causing the strange behaviour.

Reply:
We can probably overwrite the original struct, especially since the function is named `set_read_only` and takes a non-`const` pointer to it.

Reply:
If I reassign `e` (passed by ref), so:
`e = (struct employee *) cheri_perms_and(&ro_employee,  CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP)`

is the same.

Reply:
Oh, because `cheri_perms_get(&ro_employee)` should be `cheri_perms_get(ro_employee)`; you're currently testing the permissions of the pointer to the pointer.

Reply:
> Oh, because `cheri_perms_get(&ro_employee)` should be `cheri_perms_get(ro_employee)`; you're currently testing the permissions of the pointer to the pointer.

Damn copy/paste :(

Reply:
Even after removing the dangling pointer in d8e8688, the problem still persists (strlen:153).

Reply:
I think you should assume that the problem is not `strlen`.

Reply:
> I think you should assume that the problem is not `strlen`.

Yes, you are right. But those `char *` must be the problem.